### ISSUE TITLE
Cannot remove a module when it returned with flux_module_set_running 

### ISSUE THREAD

**Issue by SteVwonder** [flux-framework/flux-core#3877]
@FrankD412 and @bhatiaharsh encountered a hang in Flux during bootstrap on Summit.  They use Spack to build/install/load Flux, and the flux-sched Spack package sets `FLUX_RC_EXTRA` so that Fluxion is loaded by default when the Spack `flux-sched` package is loaded.

Unfortunately, due to a double loading bug, the `flux-sched` package is loaded twice and the `FLUX_RC_EXTRA` points to the flux-sched rc directory twice.  This leads to the flux-sched rc scripts being run twice on setup.  On the second run through, a hang occurs.  Seemingly when the resource module gets reloaded the second time (this time with qmanager also loaded).

When reproduced on a single node (either a docker container on a laptop or a single node on Summit), only the following error occurs during bootstrap:

```
2021-09-24T22:50:15.506817Z sched-fluxion-qmanager.err[0]: update_on_resource_response: exiting due to sched-fluxion-resource.notify failure: Operation canceled
```

When run on multiple nodes, a hang occurs.

Will update this issue with additional details as we collect them.

**Comment by grondo** [comment]
I think the problem is not that you "cannot remove a module when it has used `flux_module_set_running()`", but rather that a module cannot be removed before it has entered the reactor so that the `<modname>.shutdown` request from the broker can be processed.

The `flux_module_set_running()` call is provided so that a module can allow `flux module load` to return before the reactor is entered, however, the module should still enter the reactor at some point so that it can handle the builtin module services `.stats.{get,clear}` `.shutdown` `.debug` and `.ping`. I guess I'm still wondering why `sched-fluxion-resource` never enters the reactor in this case.

I think a good long term solution would be a solution to #2010 which would allow asynchronous initialization when necessary. In the case of `sched-fluxion-resource` though, initialization is "released" early so is there some reason the module can't enter the reactor instead and perform asynchronous initialization?

Even so, I wonder why `sched-fluxion-resource` is permanently blocked in a synchronous call outside the flux reactor during initialization? Shouldn't this eventually finish, the module enter the reactor, then the unload succeed?

>  But perhaps what we can do is to add a simple check into flux-core's rc1 script: for example, if FLUX_RC_EXTRA points to the base rc directory, make it a NOOP or similar.

I could be wrong, but I don't think the problem was that `FLUX_RC_EXTRA` was set to the base rc directory, but that it had two identical paths in it. We could probably try to deduplicate this list in flux-core's rc1 though.

The fluxion rc1 scripts also use `flux module reload -f` -- I wonder why that is necessary. In this case if it had been a `flux module load` we would have seen an error (presumably) and perhaps detected the problem earlier.

**Comment by dongahn** [comment]
> I think the problem is not that you "cannot remove a module when it has used flux_module_set_running()", but rather that a module cannot be removed before it has entered the reactor so that the <modname>.shutdown request from the broker can be processed.

Yes. Couldn't capture that in the short issue title. Sorry.

**Comment by dongahn** [comment]
>  guess I'm still wondering why sched-fluxion-resource never enters the reactor in this case.

> Even so, I wonder why sched-fluxion-resource is permanently blocked in a synchronous call outside the flux reactor during initialization? Shouldn't this eventually finish, the module enter the reactor, then the unload succeed?

If you want, I can do a little more debugging -- just didn't want to spend more time when this was caused with a bug in Spack recipe. 

I could be the case how `sched-fluxion-resource` is handshaking with `resource` got into some deadlock condition:
https://github.com/flux-framework/flux-sched/blob/master/resource/modules/resource_match.cpp#L1173

**Comment by dongahn** [comment]
> I could be wrong, but I don't think the problem was that FLUX_RC_EXTRA was set to the base rc directory, but that it had two identical paths in it. We could probably try to deduplicate this list in flux-core's rc1 though.

If `FLUX_RC_EXTRA` is set to the base rc directory, the rc1 stripts in it are excited twice though? I haven't looked at it closely.

**Comment by dongahn** [comment]
> The fluxion rc1 scripts also use flux module reload -f -- I wonder why that is necessary. In this case if it had been a flux module load we would have seen an error (presumably) and perhaps detected the problem earlier.

This goes way back and I think we determined `reload -f` was the only way to make our testing setting right. I will look it up when I'm available next.

**Comment by dongahn** [comment]
It is actually in the comment of the rc1 scripts:


```
# Note that we use "flux reload -f resource" instead of "flux load resource"
# to ease our testing.
# Without this, some of the in-tree tests can fail when flux-sched is already
# installed with flux-core. These installed rc1 scripts are
# executed first and disallow the "in-tree" modules from loading.
# We now use "flux reload -f resource" so that when these failing tests
# load it via FLUX_RC_EXTRA, the in-tree version can be loaded.
```

**Comment by grondo** [comment]
Thanks, sorry I missed that!

**Comment by grondo** [comment]
> If FLUX_RC_EXTRA is set to the base rc directory, the rc1 stripts in it are excited twice though? I haven't looked at it closely.
Yes, though I think the risk of this is very low since RC_EXTRA should only ever have non base directories added

**Comment by dongahn** [comment]
> Yes, though I think the risk of this is very low since RC_EXTRA should only ever have non base directories added

Yes. However, that was the original bug introduced in our Spack recipe which then had other side effects from MuMMI users that caused this hang: https://github.com/flux-framework/flux-sched/issues/861

**Comment by grondo** [comment]
> Yes. However, that was the original bug introduced in our Spack recipe which then had other side effects from MuMMI users that caused this hang

Sorry, I was probably misunderstanding your use of "base". I thought you meant "base directory" as in the location of the flux-core `rc1.d` directory. This directory is always prepended to RC_EXTRA here:

https://github.com/flux-framework/flux-core/blob/d6253f8f2da916996d0436c801147e614378ea94/etc/rc1#L34-L35

So there is no need for the flux-core spack recipe to append this RC_EXTRA. The problem the MuMMI users ran into was because the `flux-sched` package was loaded twice and it _does_ have to append to `FLUX_RC_EXTRA` (However, if we somehow deduplicated `FLUX_RC_EXTRA` it would handle either case, so a moot point I guess!)

**Comment by grondo** [comment]
It occurs to me a one line change to `flux.c` might fix this:
```diff
diff --git a/src/cmd/flux.c b/src/cmd/flux.c
index ab3bf8a02..8f7eb48b0 100644
--- a/src/cmd/flux.c
+++ b/src/cmd/flux.c
@@ -182,6 +182,7 @@ int main (int argc, char *argv[])
         environment_set_separator (env, "MANPATH", ':');
     }
 
+    environment_from_env (env, "FLUX_RC_EXTRA", "", ':');
     environment_from_env (env, "FLUX_EXEC_PATH", "", ':');
     environment_push (env, "FLUX_EXEC_PATH",
                       flux_conf_builtin_get ("exec_path", flags));
```

Because of the way `environment_from_env()` is implemented, it automatically deduplicates PATH-like environment variables (hopefully preserving order)

Before and after the change above:

```console
$ FLUX_RC_EXTRA=/foo:/foo flux /usr/bin/printenv FLUX_RC_EXTRA
/foo:/foo
$ FLUX_RC_EXTRA=/foo:/foo src/cmd/flux /usr/bin/printenv FLUX_RC_EXTRA
/foo
```
